File: src/main/java/org/roborace/lapscounter/client/WebsocketClient.java
package org.roborace.lapscounter.client;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.websocket.ClientEndpoint;
import jakarta.websocket.CloseReason;
import jakarta.websocket.ContainerProvider;
import jakarta.websocket.DeploymentException;
import jakarta.websocket.OnClose;
import jakarta.websocket.OnMessage;
import jakarta.websocket.OnOpen;
import jakarta.websocket.Session;
import jakarta.websocket.WebSocketContainer;
import org.roborace.lapscounter.domain.Message;
import org.roborace.lapscounter.domain.State;
import org.roborace.lapscounter.domain.Type;

import java.io.IOException;
import java.net.URI;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

@ClientEndpoint
public class WebsocketClient {

    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    private final String name;
    private Session userSession = null;

    private final Queue<Message> messages = new ConcurrentLinkedQueue<>();
    private Message lastMessage;
    private State state;

    public WebsocketClient(URI endpointURI, String name) throws IOException, DeploymentException {
        this.name = name;

        WebSocketContainer container = ContainerProvider.getWebSocketContainer();

        container.connectToServer(this, endpointURI);

    }

    @OnOpen
    public void onOpen(Session userSession) {
        System.out.printf("[%6s] Opening websocket userSession = [%s]\n", name, userSession);
        this.userSession = userSession;
    }

    @OnClose
    public void onClose(Session userSession, CloseReason reason) {
        System.out.printf("[%6s] Closing websocket userSession = [%s], reason = [%s]\n", name, userSession, reason);
        this.userSession = null;
    }

    @OnMessage
    public void onMessage(String message) throws IOException {
        System.out.printf("[%6s] Message received = [%s]\n", name, message);
        messages.add(OBJECT_MAPPER.readValue(message, Message.class));
    }

    public void sendMessage(String message) {
        System.out.printf("[%6s] Send message = [%s]\n", name, message);
        this.userSession.getAsyncRemote().sendText(message);
    }

    public void closeClient() {
        if (userSession != null) {
            try {
                userSession.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public Message pollMessage() {
        return messages.poll();
    }

    public boolean hasMessage() {
        return !messages.isEmpty();
    }
    public boolean hasNoMessage() {
        return messages.isEmpty();
    }

    public boolean hasMessageWithType(Type type) {
        if (messages.isEmpty()) {
            return false;
        }
        lastMessage = messages.poll();
        boolean result = lastMessage.getType() == type;
        if (result && type == Type.STATE) {
            state = lastMessage.getState();
        }
        return result;
    }

    public Message getLastMessage() {
        return lastMessage;
    }

    public String getName() {
        return name;
    }

    public State getState() {
        return state;
    }
}

File: src/main/kotlin/org/roborace/lapscounter/client/WebsocketClient.kt
package org.roborace.lapscounter.client

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import jakarta.websocket.ClientEndpoint
import jakarta.websocket.CloseReason
import jakarta.websocket.CloseReason.CloseCodes
import jakarta.websocket.ContainerProvider
import jakarta.websocket.OnClose
import jakarta.websocket.OnError
import jakarta.websocket.OnMessage
import jakarta.websocket.OnOpen
import jakarta.websocket.Session
import mu.KotlinLogging
import org.roborace.lapscounter.domain.State
import org.roborace.lapscounter.domain.Type
import org.roborace.lapscounter.domain.api.Message
import java.io.IOException
import java.net.URI
import java.util.Queue
import java.util.concurrent.ConcurrentLinkedQueue

private val logger = KotlinLogging.logger {}

@ClientEndpoint
class WebsocketClient(endpointURI: URI, val name: String) {

    private val userSession: Session = ContainerProvider.getWebSocketContainer()
        .connectToServer(this, endpointURI)

    private val messages: Queue<Message> = ConcurrentLinkedQueue()

    var lastMessage: Message = Message(type = Type.ERROR)
    var state: State? = null

    @OnOpen
    fun onOpen(userSession: Session) {
        logger.info("[{}] Opening websocket userSession = [{}]", name, userSession.id)
    }

    @OnClose
    fun onClose(userSession: Session, reason: CloseReason) {
        logger.info("[{}] Closing websocket userSession = [{}], reason = [{}]", name, userSession.id, reason)
    }

    @OnError
    fun onError(t: Throwable) {
        logger.error("[{}] Websocket error = [{}]", name, t.message)
    }

    @OnMessage
    fun onMessage(message: String) {
        logger.info("[{}] Message received = [{}]", name, message)
        messages.add(objectMapper.readValue(message, Message::class.java))
    }

    fun sendMessage(message: String) {
        logger.info("[{}] Send message = [{}]", name, message)
        userSession.asyncRemote.sendText(message)
    }

    fun closeClient() {
        logger.info("Close client $name!")
        if (userSession.isOpen) {
            try {
                userSession.close(CloseReason(CloseCodes.NORMAL_CLOSURE, "close client"))
            } catch (e: IOException) {
                logger.warn("Error closing session: ${e.message}", e)
            }
        }
    }

    fun pollMessage(): Message = messages.poll()

    fun hasMessage() = messages.isNotEmpty()

    fun hasNoMessages(): Boolean = messages.isEmpty()

    fun hasMessageWithType(type: Type): Boolean {
        lastMessage = Message(type = Type.ERROR)
        while (messages.isNotEmpty()) {
            lastMessage = messages.poll()
            if (lastMessage.type == Type.STATE) {
                state = lastMessage.state
            }
            if (lastMessage.type == type) return true
        }
        return false
    }

    companion object {
        private val objectMapper = jacksonObjectMapper()
    }
}